open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/CategoryFVT.java

```diff
@@ -45,19 +45,19 @@ public class CategoryFVT {
      *
      * Note this FVT is called by other FVTs. Who ever constructs the FVT should run deleteRemainingCategories
      */
-    private Set<String> createdCategoriesSet = new HashSet<>();
+    private final Set<String> createdCategoriesSet = new HashSet<>();
 
     public static void main(String[] args) {
         String url;
         try {
             url = RunAllFVTOn2Servers.getUrl(args);
             runWith2Servers(url);
         } catch (IOException e1) {
-            System.out.println("Error getting user input");
```

Comment:
In general we are using System.out.println across our 'interactive' applications and FVTs with the expectation of the I/O always going to console.

However we are seeing too much clutter in our build logs from FVTs, and ensuring any output generally uses logging (or audit) is the right approach. And our FVTs should really be run purely by the test runners like surefire, failsafe (for fvt)

But a main() that implements interactive tty (which is the intent here), and in a test case (vs library/production code) should probably be an exception - but only in the 'test driver' aspect ie main() or tightly related to that interaction. Indeed this part of the FVT is using a buffered input stream to query the URL if needed. Given that I think the change to log causes more obfuscation

Reply:
I hoped that the output provided by the build tool (maven/gradle) would be good enough and we could simply redirect them to log. The only thing that actually needs to change is the call to printStackTrace everything else is an incidental change so isn't important enough to change just yet.

Reply:
There is some history in the way these FVTs came about. they started as interactive applications - and so the main() allowed for the code to be run directly at the cli, or in IntelliJ. If appropriate parms were not passed, or perhaps always, the code prompted for input using an input stream. unless redirected this uses stdin, so in that context of interactive tty, using stdout seems entirely reasonable, and using log less so.

That changes when we get to the actual test case implementations as these will either be called from this code (which could control output) or most likely from a test runner - via maven/gradle. Once there I strongly agree, we should use logging, and this has big benefits to improving the quality of output from the build by hiding noise, yet making it easy to capture.

Reply:
I agree with the removal of the stack trace.

If we wanted to cleanup the FVT behaviour more, I would remove this interactive prompting for input parms & the use of input streams/stdout, probably even remove the main, and instead expect the code to always be launched from a test runner or IDE. But that would be a broader community question

Reply:
I think in general removing calls to printStackTrace is appropriate - in our runtime libraries they should be managed through our auditog framework if needed. Inadvertent disclosure is a potential security issue.

I'm less worried about FVTs in not being operational code, but it is best practice, so I would also agree with those changes. The one caveat is to be sure that if someone runs the FVTs as a java application via main() that we don't lose the exceptions completely - ie they should still be logged, or we at least report they are in the tool.. or otherwise update docs to say how they work. can comment more when the pr is ready for review

Reply:
I agree with both your points. The main thing is that we can debug an error s in the FVT when they occur in the build. We have moved to using audit logging - that writes to the console for this. I agreed with Nigel - that as a developer if I run it locally as an app to debug it outside of the build - it would be necessary to see all Exceptions.